README.md
# TWCManager

## How it works

* In order to allow multiple Tesla Wall Connector (TWC) units to operate on a single power circuit, Tesla provides a Load-Sharing protocol to interconnect the Tesla Wall Connector units together, and to direct those units to charge at a specific rate based on the number of cars charging simtultaneously.
* This script (TWCManager) pretends to be a Tesla Wall Charger (TWC) set to master mode.
* When wired to the IN or OUT pins of real TWC units set to slave mode (rotary switch position F), TWCManager can tell them to limit car charging to any whole amp value between 5A and the max rating of the charger. Charging can also be stopped so the car goes to sleep.
* This level of control is useful for having TWCManager track the real-time availability of green energy sources and direct the slave TWCs to use near the exact amount of energy available. This saves energy compared to sending the green energy off to a battery for later car charging or off to the grid where some of it is lost in transmission (and costs are greater than charging off generated solar).
* TWCManager provides built-in interfaces to several different Green Energy interfaces, to allow this tracking. Please see the table below.

## Interfaces

To enable TWCManager to gather the solar generation information that it requires to adjust the charging rate, we provide Interfaces to allow information to be retrieved (EMS Interfaces) and Interfaces to allow us to provide information to external systems to inform them of current charging output and number of cars currently charging.

### EMS Interfaces

EMS Interfaces read solar generation and/or consumption values from an external system. Current EMS interfaces are:

| Platform         | Status        | Details                 |
| ---------------- | ------------- | ----------------------- |
| Fronius Inverter | Available v1.1.0 (Soon) | Expected in next release |
| HomeAssistant    | Available v1.0.1 | Allows fetching generation and consumption from HomeAssistant sensors |
| Tesla Powerwall2 | Planned        | Expected by Mid-October 2019 |

### Status Interfaces

Status interfaces publish TWCManager status information to external systems. Current Status interfaces are:

| Platform         | Status           | Details                 |
| ---------------- | ---------------- | ----------------------- |
| HomeAssistant    | Available v1.0.1 | Provides HASS sensors to monitor TWCManager State |
| MQTT             | Available v1.0.1 | Publishes MQTT topics to monitor TWCManager State |

### Vehicle Interfaces

| Platform    | Status           | Details                 |
| ----------- | ---------------- | ----------------------- |
| TeslaAPI    | Available v1.0.0 | Allows start/stop charging via Tesla API |

## Warnings

It is important to read and understand the following warnings before using this tool:

* Misuse of the protocol described in this software can direct a Tesla Wall Charger to supply more current to a car than the charger wiring was designed for.
* This will trip a circuit breaker or may start a fire in the unlikely event that the circuit breaker fails.
* This software was not written or designed with the benefit of information from Tesla and there is always a small possibility that some unforeseen aspect of its operation could damage a Tesla vehicle or a Tesla Wall Charger.
* All efforts have been made to avoid such damage and this software is in active use on the author's own vehicle and TWC.
* In short, USE THIS SOFTWARE AT YOUR OWN RISK.

## Features
* Can be set up to only allow charging during certain hours.
* Lets you control the amount of power delivered by a Tesla Wall Connector (TWC) to the car it's charging.
This can save around 6kWh per month when used to track a local green energy source like solar panels on your roof.  It can also avoid drawing grid energy for those without net metering or limit charging to times of day when prices are cheapest.
* Whilst the TWCManager application cannot read the charge state of connected vehicles, using the Tesla API we can pull the charge status of a vehicle.
* Integrates with HomeAssistant to report TWC sensor data back to HomeAssistant for display or use in automations.

## Limitations
* Due to hardware limitations, TWCManager will not work with Tesla's older High Power Wall Connectors (HPWCs) that were discontinued around April 2016.
* Interfaces (HomeAssistant and Fronius) are in development and have not been fully tested yet. 
* There is currently no way to configure how TWCManager deals with more than one interface platform. Currently, TWCManager will add together the values (generation and consumption) for all configured integrations.

## Installation

All installation documentation (Physical and Software) is contained in the documentation repository [here](docs/README.md)

# Attribution
* Original TWCManager distribution by [cdragon](https://github.com/cdragon/TWCManager)
* Integrated improvements from [flodom's](https://github.com/flodorn/TWCManager) TWCManager fork. 
    * Support for multiple vehicles on multiple chargers
    * Support for publishing status to MQTT topics


CHANGELOG.md
# ChangeLog

This document shows the changes per release

## Current Head

  * Improvements to HomeAssistant EMS module to avoid setting Consumption/Generation values to zero as a result of a connection failure, and better exception handling in general for the module.
  * Separation of Slave TWC code from Master TWC code - this adds stability by removing a large number of global variables and reduces complexity for future feature improvements.

## v1.1.0 - Current testing branch

### New Features

  * Separation of Tesla API class from TWCManager code, this allows pluggable modules for interfacing with vehicles.
  * Added Fronius EMS module - read generation and consumption values from a Fronius inverter.

## v1.0.2 - 2019-10-11

  * Option to subtract charger utilization from consumption value if consumption meter is counting charger load.
  * Updated status output to more clearly define the generation and consumption values being tracked.

## v1.0.1 - 2019-10-06

This release is a bugfix release for v1.0.0. It contains the following fixes:

  * Fixed: MQTT status connection timeout causes a delay to heartbeat timers, which then cause TWCManager reconnections to Slaves
  * Fixed: MQTT and HASS status update rates are too high. All updates now rate limited to one per minute.

## v1.0.0 - 2019-10-05

This module provides HomeAssistant sensor functionality to the TWCManager project. It is the first release after the fork and restructuring of the code and configuration files.

### New Features

  * Forked original repository by cdragon
  * Integrated multi-charger and multi-car MQTT integration (Status Module) from flodorn's repository
  * Split the Sensor (EMS) and Status functions out of the main script
  * Created HomeAssistant EMS and Status modules and tested successfully
  
### Known Issues

  * No MQTT authentication is available currently. This will be fixed in an upcoming patch (v1.1.0).
  * Large numbers of uncaught exceptions around connectivity for EMS and Status interfaces - These can be safely ignored, they will be fixed in v1.1. Currently, they will only result in excessive logging.
  
### Future Changes

  * Split TWCMaster class for code readability
  * Separate class for serial communications - so TWCMaster and TWCSlave don't both need to speak to the serial port directly
  * Modular serial interface - allow dummy module to be substituted for testing (along with above work)


lib/TWCManager/EMS/Fronius.py
# Fronius Datamanager Solar.API Integration (Inverter Web Interface)

class Fronius:

  import requests
  import time
  
  cacheTime   = 60
  consumedW   = 0
  debugLevel  = 0
  fetchFailed = False
  generatedW  = 0
  importW     = 0
  exportW     = 0
  lastFetch   = 0
  serverIP    = None
  serverPort  = 80
  status      = False
  timeout     = 10
  voltage     = 0

  def __init__(self, debugLevel, config):
    self.debugLevel  = debugLevel
    self.status      = config.get('enabled', False)
    self.serverIP    = config.get('serverIP', None)
    self.serverPort  = config.get('serverPort','80')

  def debugLog(self, minlevel, message):
    if (self.debugLevel >= minlevel):
      print("Fronius: (" + str(minlevel) + ") " + message)

  def getConsumption(self):
    
    if (not self.status):
      self.debugLog(10, "Fronius EMS Module Disabled. Skipping getConsumption")
      return 0
    
    # Perform updates if necessary
    self.update()
    
    # Return consumption value. Fronius consumption is either negative
    # (export to grid) or positive (import from grid). We add generation
    # value to make it the delta between this and current consumption
    if ((self.consumedW < 0) or (self.consumedW > 0)):
      return float(self.generatedW + self.consumedW)
    else:
      return float(0)

  def getGeneration(self):
    
    if (not self.status):
      self.debugLog(10, "Fronius EMS Module Disabled. Skipping getGeneration")
      return 0

    # Perform updates if necessary
    self.update()

    # Return generation value
    return float(self.generatedW)
  
  def getInverterData(self):

    url = "http://" + self.serverIP + ":" + self.serverPort
    url = url + "/solar_api/v1/GetInverterRealtimeData.cgi?Scope=Device&DeviceID=1&DataCollection=CommonInverterData"

    return self.getInverterValue(url)
  
  def getInverterValue(self, url):
    
    # Fetch the specified URL from the Fronius Inverter and return the data
    self.fetchFailed = False
    
    try:
        r = self.requests.get(url, timeout=self.timeout)
    except self.requests.exceptions.ConnectionError as e: 
        self.debugLog(4, "Error connecting to Fronius Inverter to fetch sensor value")
        self.debugLog(10, str(e))
        self.fetchFailed = True
        return False
      
    r.raise_for_status()
    jsondata = r.json()
    return jsondata

  def getMeterData(self):
    url = "http://" + self.serverIP + ":" + self.serverPort
    url = url + "/solar_api/v1/GetMeterRealtimeData.cgi?Scope=Device&DeviceId=0"

    return self.getInverterValue(url)

  def update(self):
    
    if ((int(self.time.time()) - self.lastFetch) > self.cacheTime):
      # Cache has expired. Fetch values from HomeAssistant sensor.
      
      inverterData = self.getInverterData()
      if (inverterData):
        try:
          self.generatedW = inverterData['Body']['Data']['PAC']['Value']
        except (KeyError, TypeError) as e:
          self.debugLog(4, "Exception during parsing Inveter Data (PAC)")
          self.debugLog(10, e)

        try:
          self.voltage = inverterData['Body']['Data']['UAC']['Value']
        except (KeyError, TypeError) as e:
          self.debugLog(4, "Exception during parsing Inveter Data (UAC)")
          self.debugLog(10, e)

      meterData = self.getMeterData()
      if (meterData):
        try:
          self.consumedW = float(meterData['Body']['Data']['PowerReal_P_Sum'])
        except (KeyError, TypeError) as e:
          self.debugLog(4, "Exception during parsing Meter Data (Consumption)")
          self.debugLog(10, e)
    
      # Update last fetch time
      if (self.fetchFailed is not True):
        self.lastFetch = int(self.time.time())
        
      return True
    else:
      # Cache time has not elapsed since last fetch, serve from cache.
      return False


lib/TWCManager/Status/HASSStatus.py
# HomeAssistant Status Output
# Publishes the provided sensor key and value pair to a HomeAssistant instance

class HASSStatus:

  import time
  import requests
  
  apiKey           = None
  debugLevel       = 0
  msgRate          = {}
  msgRatePerSensor = 60
  status           = False
  serverIP         = None
  serverPort       = 8123
  timeout          = 2
  
  def __init__(self, debugLevel, config):
    self.status      = config.get('enabled', False)
    self.serverIP    = config.get('serverIP', None)
    self.serverPort  = config.get('serverPort', 8123)
    self.apiKey      = config.get('apiKey', None)
    self.debugLevel  = debugLevel

  def debugLog(self, minlevel, message):
    if (self.debugLevel >= minlevel):
      print("debugLog: (" + str(minlevel) + ") " + message)

  def setStatus(self, twcid, key, value):
    sensor = "sensor.twcmanager_" + str(twcid.decode("utf-8")) + "_" + key

    if (self.status):

      # Perform rate limiting first (as there are some very chatty topics).
      # For each message that comes through, we take the sensor name and check
      # when we last updated it. If it was less than msgRatePerSensor
      # seconds ago, we dampen it.
      if (sensor in self.msgRate):
        if ((self.time.time() - self.msgRate[sensor]) < self.msgRatePerSensor):
          return True
        else:
          self.msgRate[sensor] = self.time.time()
      else:
        self.msgRate[sensor] = self.time.time()

      url = "http://" + self.serverIP + ":" + self.serverPort 
      url = url + "/api/states/" + sensor
      headers = {
          'Authorization': 'Bearer ' + self.apiKey,
          'content-type': 'application/json'
      }

      try:
        self.debugLog(8, "Sending POST request to HomeAssistant for sensor " + sensor + "(value " + str(value) + ").")
        self.requests.post(url, json={"state":value}, timeout=self.timeout, headers=headers)
      except self.requests.exceptions.ConnectionError as e:
        self.debugLog(4, "Error connecting to HomeAssistant to publish sensor values")
        self.debugLog(10, str(e))
        return False
      except self.requests.exceptions.ReadTimeout as e:
        self.debugLog(4, "Error connecting to HomeAssistant to publish sensor values")
        self.debugLog(10, str(e))
        return False
      except Exception as e:
        self.debugLog(4, "Error during publishing HomeAssistant sensor values")
        self.debugLog(10, str(e))
        return False


lib/TWCManager/Status/MQTTStatus.py
# MQTT Status Output
# Publishes the provided key and value pair to the provided topic prefix

class MQTTStatus:

  import time
  import paho.mqtt.client as mqtt

  brokerIP        = None
  brokerPort      = 1883
  connectionState = 0
  debugLevel      = 0
  msgQueue        = []
  msgQueueBuffer  = []
  msgQueueMax     = 16
  msgRate         = {}
  msgRatePerTopic = 60
  password        = None
  status          = False
  serverTLS       = None
  topicPrefix     = None
  username        = None

  def __init__(self, debugLevel, config):
    self.debugLevel  = debugLevel
    self.status      = config.get('enabled', False)
    self.brokerIP    = config.get('brokerIP', None)
    self.topicPrefix = config.get('topicPrefix', None)
    self.username    = config.get('username', None)
    self.password    = config.get('password', None)

  def debugLog(self, minlevel, message):
    if (self.debugLevel >= minlevel):
      print("MQTTStatus: (" + str(minlevel) + ") " + message)

  def setStatus(self, twcid, key, value):
    if (self.status):
      topic = self.topicPrefix+ "/" + str(twcid.decode("utf-8"))
      topic = topic + "/" + key

      # Perform rate limiting first (as there are some very chatty topics).
      # For each message that comes through, we take the topic name and check
      # when we last sent a message. If it was less than msgRatePerTopic
      # seconds ago, we dampen it.
      if (topic in self.msgRate): 
        if ((self.time.time() - self.msgRate[topic]) < self.msgRatePerTopic):
          return True
        else:
          self.msgRate[topic] = self.time.time()
      else:
        self.msgRate[topic] = self.time.time()

      # Now, we push the message that we'd like to send into the
      # list of messages to be published once a connection is established
      msg = { "topic": topic, "payload": value }
      self.msgQueue.append(msg.copy())

      # If msgQueue size exceeds msgQueueMax, trim the list size
      # This will discard MQTT messages in environments where the MQTT
      # broker cannot accept the rate of MQTT messages sent
      if (len(self.msgQueue) > (self.msgQueueMax + 8)):
        del self.msgQueue[:self.msgQueueMax]

      # Now, we attempt to establish a connection to the MQTT broker
      if (self.connectionState == 0):
        self.debugLog(10, "MQTT Status: Attempting to Connect")
        try:
          client = self.mqtt.Client()
          if (self.username and self.password):
            client.username_pw_set(self.username, self.password)
          client.on_connect = self.mqttConnected
          client.connect_async(self.brokerIP, port=self.brokerPort, keepalive=30)
          self.connectionState = 1
          client.loop_start()
        except ConnectionRefusedError as e:
          self.debugLog(4, "Error connecting to MQTT Broker to publish topic values")
          self.debugLog(10, str(e))
          return False
        except OSError as e:
          self.debugLog(4, "Error connecting to MQTT Broker to publish topic values")
          self.debugLog(10, str(e))
          return False

  def mqttConnected(self, client, userdata, flags, rc):
    # This callback function is called once the MQTT client successfully
    # connects to the MQTT server. It will then publish all queued messages
    # to the server, and then disconnect.

    self.debugLog(10, "Connected to MQTT Broker with RC: " + str(rc))
    self.debugLog(11, "Copy Message Buffer")
    self.msgQueueBuffer = self.msgQueue.copy()
    self.debugLog(11, "Clear Message Buffer")
    self.msgQueue.clear()

    for msg in self.msgQueueBuffer:
      self.debugLog(8, "Publishing MQTT Topic " + str(msg['topic']))
      try:
        client.publish(msg['topic'], payload=msg['payload'])
      except e:
        self.debugLog(4, "Error publishing MQTT Topic Status")
        self.debugLog(10, str(e))

    client.loop_stop()
    self.msgQueueBuffer.clear()
    self.connectionState = 0
    client.disconnect()


lib/TWCManager/EMS/HASS.py
class HASS:

  # HomeAssistant EMS Module
  # Fetches Consumption and Generation details from HomeAssistant
  
  import requests
  import time
  
  apiKey                = None
  cacheTime             = 60
  consumedW             = 0
  debugLevel            = 0
  fetchFailed           = False
  generatedW            = 0
  hassEntityConsumption = None
  hassEntityGeneration  = None
  lastFetch             = 0
  status                = False
  serverIP              = None
  serverPort            = 8123
  timeout               = 10
  
  def __init__(self, debugLevel, config):
    self.status                = config.get('enabled', False)
    self.serverIP              = config.get('serverIP', None)
    self.serverPort            = config.get('serverPort', 8123)
    self.apiKey                = config.get('apiKey', None)
    self.debugLevel            = debugLevel
    self.hassEntityConsumption = config.get('hassEntityConsumption', None)
    self.hassEntityGeneration  = config.get('hassEntityGeneration', None)

  def debugLog(self, minlevel, message):
    if (self.debugLevel >= minlevel):
      print("debugLog: (" + str(minlevel) + ") " + message)
      
  def getConsumption(self):
    
    if (not self.status):
      self.debugLog(10, "HASS EMS Module Disabled. Skipping getConsumption")
      return 0
    
    # Perform updates if necessary
    self.update()
    
    # Return consumption value
    return self.consumedW

  def getGeneration(self):
    
    if (not self.status):
      self.debugLog(10, "HASS EMS Module Disabled. Skipping getConsumption")
      return 0
    
    # Perform updates if necessary
    self.update()
    
    # Return generation value
    return self.generatedW
  
  def getAPIValue(self, entity):
    url = "http://" + self.serverIP + ":" + self.serverPort + "/api/states/" + entity
    headers = {
        'Authorization': 'Bearer ' + self.apiKey,
        'content-type': 'application/json'
    }

    # Update fetchFailed boolean to False before fetch attempt
    # This will change to true if the fetch failed, ensuring we don't then use the value to update our cache
    self.fetchFailed = False
    
    try:
        self.debugLog(10, "Fetching HomeAssistant EMS sensor value " + str(entity))
        httpResponse = self.requests.get(url, headers=headers, timeout=self.timeout)
    except self.requests.exceptions.ConnectionError as e: 
        self.debugLog(4, "Error connecting to HomeAssistant to fetch sensor value")
        self.debugLog(10, str(e))
        self.fetchFailed = True
        return False
    except self.requests.exceptions.ReadTimeout as e:
        self.debugLog(4, "Read Timeout occurred fetching HomeAssistant sensor value")
        self.debugLog(10, str(e))
        self.fetchFailed = True
        return False

    jsonResponse = httpResponse.json() if httpResponse and httpResponse.status_code == 200 else None

    if jsonResponse:
        return jsonResponse["state"]
    else:
        return None

  def setCacheTime(self, cacheTime):
    self.cacheTime = cacheTime
    
  def setTimeout(self, timeout):
    self.timeout = timeout
    
  def update(self):
    # Update function - determine if an update is required
    
    if ((int(self.time.time()) - self.lastFetch) > self.cacheTime):
      # Cache has expired. Fetch values from HomeAssistant sensor.
            
      if (self.hassEntityConsumption):
          apivalue = self.getAPIValue(self.hassEntityConsumption)
          if (self.fetchFailed is not True):
              self.debugLog(10, "HASS getConsumption returns " + str(apivalue))
              self.consumedW = float(apivalue)
          else:
              self.debugLog(10, "HASS getConsumption fetch failed, using cached values")
      else:
          self.debugLog(10, "HASS Consumption Entity Not Supplied. Not Querying")

      if (self.hassEntityGeneration):
          apivalue = self.getAPIValue(self.hassEntityGeneration)
          if (self.fetchFailed is not True):
              self.debugLog(10, "HASS getGeneration returns " + str(apivalue))
              self.generatedW = float(apivalue)
          else:
              self.debugLog(10, "HASS getGeneration fetch failed, using cached values")
      else:
          self.debugLog(10, "HASS Generation Entity Not Supplied. Not Querying")

      # Update last fetch time
      if (self.fetchFailed is not True):
        self.lastFetch = int(self.time.time())
        
      return True
    else:
      # Cache time has not elapsed since last fetch, serve from cache.
      return False
